tf2/src/buffer_core.cpp

```diff
@@ -1283,6 +1283,23 @@ TransformableRequestHandle BufferCore::addTransformableRequest(
   const std::string & source_frame,
   TimePoint time)
 {
+  // Even though we only modify transformable_requests_ at the end of the
+  // method, we still need to take the lock at the beginning.  This is to
+  // ensure that we don't have a TOCTTOU race between this method and
+  // testTransformableRequests.  If the lock were only at the end of this
+  // method, the race occurs like this:
+  //
+  // T1: addTransformableRequest, determines that needs to add to transformable_requests_
+  // T2: in testTransformableRequests already, holding the lock
+  // T1: blocked getting lock
+  // T2: calls all callbacks for outstanding transforms (doesn't include the current one)
+  // T2: unlocks
+  // T1: gets lock, adds to list
+  //
+  // If nothing ever calls setTransform() again, then the callback for the
+  // current request will never get called.  We fix this by holding the mutex
+  // across the entirety of this call.
+  std::unique_lock<std::mutex> lock(transformable_requests_mutex_);
```

Comment:
It should be possible to move the lock a bit down within the function. The first condition/return should be obvious to skip but I would think it could even go down as far as https://github.com/ros2/geometry2/pull/307/files#diff-23f1780342d49d752e5fabba0fa7de1cR1327

Reply:
Yeah, totally agreed with you about moving it after the `if (target_frame == source_frame)` ; that should make it cheaper in a probably-common case.  I'll do that.

But I don't think I can move it below the check for `canTransformInternal`.  That's exactly the race; if `canTransformInternal` returns false, then we've already decided it needs to be added to the callback list.  If `testTransformableRequests` has the lock and is executing at that time, then the callback won't yet be on the list, so it won't get called there.  Since we've determined it isn't an "immediate" case the callback won't get called in `addTransformableRequest` either.  So we'll have essentially lost the event.

Reply:
Not sure how frequent this is being called. An alternative would be to check `canTransformInternal` without locking the mutex to allow for the early exit. And afterwards while holding the lock check `canTransformInternal` again to ensure to avoid the TOCTTOU. Just a thought though.

Reply:
I think that calling `canTransformInternal` twice is going to be more expensive than taking a lock.  For now, I'm going to leave this code as-is; if this turns out to be expensive later on, we can revisit the architecture and see if we can make it faster.